pkg/controller/controller_test.go

```diff
@@ -664,13 +665,36 @@ func (s *ControllerSuite) TestDeferPhase(c *C) {
 	err = s.waitOnActionSetState(c, as, crv1alpha1.StateRunning)
 	c.Assert(err, IsNil)
 
+	// we need set and not string becauce we are continuously getting runningPhase and storing
+	// it, so value would be duplicate, we don't want duplicates
+	runningPhases := sets.NewString()
+	go func() {
+		// get actionset and store the runningPhase fields to a set
+		for {
+			a, err := s.crCli.ActionSets(s.namespace).Get(ctx, as.Name, metav1.GetOptions{})
+			c.Assert(err, IsNil)
+			if a.Status.Progress.RunningPhase != "" {
+				runningPhases.Insert(a.Status.Progress.RunningPhase)
+			}
+		}
+	}()
+
 	// make sure deferPhase is also run successfully
 	err = s.waitOnDeferPhaseState(c, as, crv1alpha1.StateComplete)
 	c.Assert(err, IsNil)
 
 	err = s.waitOnActionSetState(c, as, crv1alpha1.StateComplete)
 	c.Assert(err, IsNil)
 
+	// once actionset is completed make sure all the phases are in runningPhases
+	// that would confirm that the actionset's status.progress.runningPhase was set
+	// to those phases
+	op := sets.NewString()
+	// these phases are from blueprint that we create above
+	// and after actionset completion the runningPhase becomes ""
+	op.Insert("backupPhaseOne").Insert("backupPhaseTwo").Insert("deferPhase")
```

Comment:
I think didn't understand it well. Even if we capture the running phase in `waitOnActionSetState` we will do the same right? We would get an actionset and see the running phase. If the phases changes from a to b and then from b to c again within one `get` call, we would not get `b` in our test but this would be the, I think, even if we put this in `waitOnActionSetState`.
What do you think?

Reply:
That's correct. Capturing it in `waitOnActionSetState` doesn't guarantee a successful run. We have faced a race in that function too whenever the execution was fast. It would timeout waiting for the `running` state while the action set had already reached the `completed` state.
But since we are already getting the action set to check status, the failure chances would be the same as the state failure.

Reply:
mmm, yeah. I think we can somehow achieve that. 
But I think it's not that straight forward. For example if we try store the running phases in this line
```
err = s.waitOnActionSetState(c, as, crv1alpha1.StateRunning)
```
we just wait for the action to be running. And after that we return. So we will not have all the phases stored in the `runningPhase` var.

If we are talking about storing the phases in this line 
```
err = s.waitOnActionSetState(c, as, crv1alpha1.StateComplete)
```
In that case there are chances that some of the phases are already run so we will not be able to capture those phases.

Should I try this?

Reply:
Yeah, that seems problematic too. Let's merge and see if it flakes 🚀

Reply:
thank you.